bam_consensus.c

```diff
@@ -0,0 +1,1522 @@
+}
+
+/* --------------------------------------------------------------------------
+ * A bayesian consensus algorithm that analyses the data to work out
+ * which hypothesis of pure A/C/G/T/absent and all combinations of two
+ * such bases meets the observations.
+ *
+ * This has its origins in Gap4 (homozygous) -> Gap5 (heterozygous)
+ * -> Crumble (tidied up to use htslib's pileup) -> here.
+ *
+ */
+
+#define CONS_DISCREP    4
+#define CONS_ALL        15
+
+#define CONS_MQUAL      16
+
+// Make these a parameter, but capping mqual seems wise as it's
+// not so robust especially at the low end when real SNPs exist,
+// but reduce mqual for many algorithms.
+#define MIN_MQUAL       10
+#define MAX_MQUAL       50
+
+typedef struct {
+    /* the most likely base call - we never call N here */
+    /* A=0, C=1, G=2, T=3, *=4 */
+    int call;
+
+    /* The most likely heterozygous base call */
+    /* Use "ACGT*"[het / 5] vs "ACGT*"[het % 5] for the combination */
+    int het_call;
+
+    /* Log-odds for het_call */
+    int het_logodd;
+
+    /* Single phred style call */
+    int phred;
+
+    /* Sequence depth */
+    int depth;
+
+    /* Discrepancy search score */
+    float discrep;
+
+    /* Base modifications, for top (0) and bottom (1) strands */
+    int mod_call[2];  // +ve for char, -ve for ChEBI ID
+    int mod_qual[2];  // phred scale
+} consensus_t;
+
+/*
+ * Produce the insertion portion of the consensus sequence.
+ * Returns 0 on success, filling out ins_ks and ins_ks,
+ *        -1 on failure
+ */
+static int insertion_consensus(const bam_pileup1_t *plp, int nplp,
+                               consensus_opts *opts, int nins, int ins_len,
+                               kstring_t *ins_ks, kstring_t *inq_ks) {
+    // Map "seqi" nt16 to A,C,G,T compatibility with weights on pure bases.
+    // where seqi is A | (C<<1) | (G<<2) | (T<<3)
+    //                        * A C M  G R S V  T W Y H  K D B N
+    static int seqi2A[16] = { 0,8,0,4, 0,4,0,2, 0,4,0,2, 0,2,0,1 };
+    static int seqi2C[16] = { 0,0,8,4, 0,0,4,2, 0,0,4,2, 0,0,2,1 };
+    static int seqi2G[16] = { 0,0,0,0, 8,4,4,1, 0,0,0,0, 4,2,2,1 };
+    static int seqi2T[16] = { 0,0,0,0, 0,0,0,0, 8,4,4,2, 8,2,2,1 };
+
+    int ins_[128][16], (*ins)[16] = ins_;
+    int i;
+
+    if (ins_len > 128)
+        if (!(ins = calloc(ins_len * 16, sizeof(int))))
+            return -1;
+
+    // Accumulate into ins[][] arrays
+    kstring_t is = {0,0};
+    for (i = 0; i < nplp; i++) {
+        const bam_pileup1_t *p = plp+i;
+
+        // Slow method for now is bam_plp_insertion and parse back.
+        // NB no support for opts->use_qual yet here.  We either need
+        // a better bam_plp_insertion or inline it here with added
+        // quality value processing.
+        int il = bam_plp_insertion(p, &is, NULL), j;
+        if (il < ins_len)
+            continue;
+
+        for (j = 0; j < il; j++) {
+            if (is.s[j] == '*') {
+                ins[j][16] += 8;
```

Comment:
There's a compiler error on this line:
```
bam_consensus.c:210:28: error: array subscript 16 is above array bounds of 'int[16]' [-Werror=array-bounds]
  210 |                 ins[j][16] += 8;
      |                            ^~
```

Should the array size be 17, or should this be index 0?  (Changing to 0 would also require adjustments below).